pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -564,77 +562,59 @@ private void doAddConsumingSegment(String segmentName)
-          PeerServerSegmentFinder.getOnlineServersFromExternalView(_helixManager.getClusterManagmentTool(),
-              _helixManager.getClusterName(), _tableNameWithType, zkMetadata.getSegmentName(), _peerDownloadScheme,
-              onlineServerURIs);
-          if (!onlineServerURIs.isEmpty()) {
-            return downloadSegmentFromPeers(zkMetadata);
-          }
-        } catch (Exception e) {
-          _logger.warn("Could not download segment: {} from peer", zkMetadata.getSegmentName(), e);
-        }
+      zkMetadata = fetchZKMetadata(segmentName);
+      if (zkMetadata.getStatus().isCompleted()) {
+        return super.downloadSegment(zkMetadata);
       }
 
-      long timeElapsed = System.currentTimeMillis() - startTime;
-      long timeRemaining = downloadTimeoutMilliseconds - timeElapsed;
-
-      if (timeRemaining <= 0) {
+      long timeRemainingMs = deadlineMs - System.currentTimeMillis();
+      if (timeRemainingMs <= 0) {
         break;
       }
 
-      _logger.info("Sleeping for 30 seconds as the segment url is missing. Time remaining: {} minutes",
-          Math.round(timeRemaining / 60000.0));
-
-      // Sleep for the shorter of our normal interval or remaining time
-      Thread.sleep(Math.min(SLEEP_INTERVAL_MS, timeRemaining));
+      long sleepTimeMs = Math.min(SLEEP_INTERVAL_MS, timeRemainingMs);
+      _logger.info("Sleeping for: {}ms waiting for segment: {} to be completed. Time remaining: {}ms", sleepTimeMs,
+          segmentName, timeRemainingMs);
+      //noinspection BusyWait
+      Thread.sleep(sleepTimeMs);
     }
 
     // If we exit the loop without returning, throw an exception
     throw new TimeoutException(
-        "Failed to download segment after " + TimeUnit.MILLISECONDS.toMinutes(downloadTimeoutMilliseconds)
-            + " minutes of retrying. Segment: " + zkMetadata.getSegmentName());
-  }
-
-  private long getDownloadTimeOutMilliseconds(@Nullable TableConfig tableConfig) {
-    return Optional.ofNullable(tableConfig).map(TableConfig::getIngestionConfig)
-        .map(IngestionConfig::getStreamIngestionConfig).map(StreamIngestionConfig::getStreamConfigMaps)
-        .filter(maps -> !maps.isEmpty()).map(maps -> maps.get(0)).map(map -> map.get(SEGMENT_DOWNLOAD_TIMEOUT_MINUTES))
-        .map(timeoutStr -> {
-          try {
-            return TimeUnit.MINUTES.toMillis(Long.parseLong(timeoutStr));
-          } catch (NumberFormatException e) {
-            return DEFAULT_SEGMENT_DOWNLOAD_TIMEOUT_MS;
-          }
-        }).orElse(DEFAULT_SEGMENT_DOWNLOAD_TIMEOUT_MS);
+        "Failed to download segment: " + segmentName + " after: " + downloadTimeoutMs + "ms of retrying");
+  }
+
+  private long getDownloadTimeoutMs(TableConfig tableConfig) {
+    Map<String, String> streamConfigMap = IngestionConfigUtils.getStreamConfigMaps(tableConfig).get(0);
+    String timeoutSeconds = streamConfigMap.get(StreamConfigProperties.PAUSELESS_SEGMENT_DOWNLOAD_TIMEOUT_SECONDS);
```

Comment:
Except for backward compatibility, we are unnecessarily keeping time units different for these timeouts
Let's try to keep both in seconds or minutes